### PR TITLE
Allow window elements as well as iframes to the the target of mozbrowser events

### LINKED ISSUE #12420: Allow mozbrowser events to be targeted at the window
#11248 requires throwing a mozbrowsererror event at the root window rather than at an iframe (because we may not know which iframe caused the error). At the moment, mozbrowser events can only be targeted at iframes, we should fix this so they can target other elements.

### PR DESCRIPTION
**Pull request by asajeffrey**
Allow mozbrowser events, in particular mozbrowsererror events, to target a window. Needed for https://github.com/browserhtml/browserhtml/issues/1182

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #12420
- [X] These changes do not require tests because we're not testing our issue reporting system, which this is intended for.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12426)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @Ms2ger @Gozala @paulrouget

File: components/script/dom/htmliframeelement.rs [hunk 1]
```diff
@@ -335,97 +321,104 @@ impl HTMLIFrameElementLayoutMethods for LayoutJS<HTMLIFrameElement> {
...
 }
 
-pub trait MozBrowserEventDetailBuilder {
-    #[allow(unsafe_code)]
-    unsafe fn build_mozbrowser_event_detail(&self,
-                                            event: MozBrowserEvent,
-                                            cx: *mut JSContext,
-                                            rval: MutableHandleValue);
+#[allow(unsafe_code)]
+pub unsafe fn build_mozbrowser_custom_event(window: Root<Window>, event: MozBrowserEvent) -> Root<CustomEvent> {
```
Comment (Ms2ger) [inline components/script/dom/htmliframeelement.rs]: This function provides a safe abstraction; it should just have an internal `unsafe {}` block.

Also, arguments should *never* be `Root<T>`; use `&Window`.

File: components/script_traits/lib.rs [hunk 2]
```diff
@@ -177,7 +177,7 @@ pub enum ConstellationControlMsg {
     /// Notifies script thread that a url should be loaded in this iframe.
     Navigate(PipelineId, SubpageId, LoadData),
     /// Requests the script thread forward a mozbrowser event to an iframe it owns
```
Comment (Ms2ger) [inline components/script_traits/lib.rs]: Make sure to update all the doc comments for things you change to explain what happens if the subpage id is none.

File: components/script/dom/htmliframeelement.rs (hunk 1 above)
Comment (asajeffrey) [inline components/script/dom/htmliframeelement.rs]: Fixed.

File: components/script_traits/lib.rs (hunk 2 above)
Comment (asajeffrey) [inline components/script_traits/lib.rs]: Fixed.

**Comment by asajeffrey** [comment]
Rebased.

**Comment by SimonSapin** [comment]
@bors-servo r+

---

Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12426#-:-KN6HJltHE1OqxUsA_h0:1236647940)*

**Pull request merged**